lib/bap_disasm/bap_disasm_x86_lifter.ml

```diff
@@ -0,0 +1,2319 @@
+    let open Exp in
+    let cond = match signed with
+      | true -> e1 <$ e2
+      | false -> e1 < e2 in
+    ite (lnot cond) e1 e2
+
+  let (<<%) = BZ.shift_left_big_int
+  let (+%) = BZ.add_big_int
+  let bi1 = BZ.big_int_of_int 0x1
+  let power_of_two = BZ.shift_left_big_int bi1
+  let bitmask = let (-%) = BZ.sub_big_int in
+    (fun i -> power_of_two i -% bi1)
+  let to_big_int (i,t) =
+    let bits = bits_of_width t in
+    BZ.and_big_int i (bitmask bits)
+  (* sign extend to type t *)
+  let to_sbig_int (i,t) =
+    let (>>%) = BZ.shift_right_big_int in
+    let (-%) = BZ.sub_big_int in
+    let bi_is_zero bi = BZ.(big_int_of_int 0x0 |> eq_big_int bi) in
+    let bits = bits_of_width t in
+    let final = to_big_int (i, Type.imm (bits-1)) in
+    (* mod always returns a positive number *)
+    let sign = i >>% (bits-1) in
+    if bi_is_zero sign then
+      (* positive *) final
+    else (* negative *) BZ.minus_big_int ((power_of_two (bits-1) -% final))
+  let to_signed i t = to_sbig_int (i, t)
+  let to_val t i =
+    (to_big_int (i,t), t)
+  let cast ct ((_,t) as v) t2 =
+    let bits1 = bits_of_width t in
+    let bits = bits_of_width t2 in
+    let open Exp.Cast in
+    match ct with
+    | UNSIGNED -> to_val t2 (to_big_int v)
+    | SIGNED -> to_val t2 (to_sbig_int v)
+    | HIGH -> to_val t2 (BZ.shift_right_big_int (to_big_int v) (bits1-bits))
+    | LOW -> to_val t2 (to_big_int v)
+end
+open Big_int_temp
+
+module Cpu_exceptions = struct
+  let general_protection = Stmt.cpuexn 0xd
+  let divide_by_zero = Stmt.cpuexn 0x0
+end
+
+
+let compute_segment_bases = ref false
+
+(* Note: In general, the function g is the get memory function.  The
+   variable na refers to the next address or next instruction.
+
+   To help understand this file, please refer to the Intel Instruction
+   Set Reference. For consistency, any section numbers here are wrt
+   Order Number: 253666-035US June 2010 and 253667-035US.
+
+
+   The x86 instruction format is as follows:
+   Instruction Prefixes: 0-4bytes (1 byte per prefix)
+   Optional Rex Prefix: 1 byte
+   Opcode: 1 - 3 bytes.
+   ModR/M: 1 optional byte
+   SIB: 1 optional byte
+   Displacement: 0,1,2, or 4 bytes.
+   Immediate: 0,1,2, or 4 bytes
+
+   ModR/M has the following format:
+   7:6 Mod
+   5:3 Reg or extra opcode bits
+   2:0 R/M
+
+   SIB:
+   7:6 Scale
+   5:3 Index
+   2:0 Base
+
+
+   In order to get the most common unsupported opcodes, you can run something like:
+   for f in bin/*; do BAP_DEBUG_MODULES=AsmirV ~/bap/trunk/utils/iltrans -bin $f ; done 2>&1  >/dev/null  | grep opcode | sed 's/.*opcode: //' | sort | uniq -c | sort -n
+
+   To optimize for number of programs disassembled:
+   for f in bin/*; do echo -n "$f "; BAP_DEBUG_MODULES=AsmirV iltrans -bin $f 2>&1  >/dev/null  | grep opcode | sed 's/.*opcode: //' | sort | uniq -c | sort -n  | wc -l; done | sort -n -k 2
+
+*)
+
+(* type segment = CS | SS | DS | ES | FS | GS *)
+
+type binopf = Exp.binop
```

Comment:
It would be a good idea to move all type definitions into a separate module, named, for example, `Bap_disasm_x86_types` and move all variable definitions, like registers, cpu exceptions, and all other platform specific into `Bap_disasm_x86_env` module. C.f., corresponding ARM modules.